R/datasets_api_client.R

```diff
@@ -0,0 +1,216 @@
+
+      self$api_client$stop_for_status(response)
+      httr::content(response, as = "parsed")
+    },
+    write_file = function(
+        dataset_rid,
+        file_path,
+        branch_id = NULL,
+        transaction_type = NULL,
+        transaction_rid = NULL,
+        body = NULL) {
+      url_path <- sprintf("/%s/files:upload", dataset_rid)
+      header_params <- c("Content-Type" = "application/octet-stream")
+      query_params <- list(
+        "preview" = TRUE
+      )
+      query_params["filePath"] <- file_path
+      query_params["branchId"] <- branch_id
+      query_params["transactionType"] <- transaction_type
+      query_params["transactionRid"] <- transaction_rid
+
+      response <- self$api_client$call_api(
+        url = paste0(self$api_client$base_path, url_path),
+        method = "POST",
+        query_params = query_params,
+        header_params = header_params,
+        body = body)
+
+      self$api_client$stop_for_status(response)
+      httr::content(response, as = "parsed")
+    },
+    read_file = function(
+        dataset_rid,
+        file_path,
+        branch_id = NULL,
+        start_transaction_rid = NULL,
+        end_transaction_rid = NULL) {
+      url_path <- sprintf("/%s/files/%s/content", dataset_rid, utils::URLencode(file_path, reserved = TRUE))
+      query_params <- list(
+        "preview" = TRUE
+      )
+      query_params["branchId"] <- branch_id
+      query_params["startTransactionRid"] <- start_transaction_rid
+      query_params["endTransactionRid"] <- end_transaction_rid
+
+      response <- self$api_client$call_api(
+        url = paste0(self$api_client$base_path, url_path),
+        method = "GET",
+        query_params = query_params)
+
+      self$api_client$stop_for_status(response)
+    },
+    # Internal
+    download_files = function(alias, file_paths) {
+      url_path <- sprintf("/%s/files/download", dataset_rid)
+      body <- jsonlite::toJSON(list(files = file_paths))
+      response <- self$api_client$call_api(
+        url = paste0(self$api_client$base_path, url_path),
+        method = "POST",
+        body = body)
+
+      self$api_client$stop_for_status(response)
+    },
+    write_file_internal = function(alias, file_path, body = NULL) {
+      url_path <- sprintf("/%s/files/%s/content", dataset_rid, utils::URLencode(file_path, reserved = TRUE))
+      query_params <- list(
+        "preview" = TRUE
+      )
+
+      response <- self$api_client$call_api(
+        url = paste0(self$api_client$base_path, url_path),
+        method = "PUT",
+        query_params = query_params,
+        header_params = header_params,
+        body = body)
+
+      self$api_client$stop_for_status(response)
+      httr::content(response, as = "parsed")
+    }
+  ),
+)
+
+#' @keywords internal
+get_datasets_client <- function() {
+  base_path <- get_config("datasets.url", NULL)
+  if (is.null(base_path)) {
+    base_path <- paste0("https://", get_config("hostname"), "/api/v1/datasets")
```

Comment:
we may want some sanity checking here to make sure people only provide the hostname? feels like this could easily lead to spurious errors

Reply:
we should agree with the python team on this so we have a consistent experience